### DISCUSSION TITLE
Arquero vs Vega-Lite transform methods?

### DISCUSSION THREAD

**Discussion by gordonje** [uwdata/arquero#243] [Q&A] [answered]
Arquero with the Vega Lite API on Observable has been my primary toolchain for most of the last year. So first off, thank you again for your amazing software packages and resources!

One part of this set up I've been struggling with a lately is when to use Arquero versus the transform methods in the Vega Lite API. All things being equal: It feels easier to do all of my transformations with Arquero, then pass the table object when calling [`vl.data()`](https://vega.github.io/vega-lite-api/api/data). This seems like the path of least resistance especially in cases when I'm allowing user input (via [Observable's inputs](https://observablehq.com/@observablehq/inputs) or other cells defined with the `viewof` operator) that dynamically filter the table rows and map columns to encoding channels.

Are there advantages to using Vega Lite's transform methods? Your notebook re: [using Arquero with D3](https://observablehq.com/@uwdata/arquero-and-d3) was helpful, and I wonder if you have similar guidance for using Arquero with Vega Lite.

I should also add: I usually fetch data from an S3 bucket with restricted access. I use the AWS SDK to fetch the data, then `aq.fromCSV()` or `aq.fromArrow()` to start processing it. So can't just pass a URL to `vl.data()` and have it fetch for me (right?)

Thank you,

**Comment by jheer** [comment, marked as answer]
Ultimately the answer depends on your use case and your desired level of portability. If you want data transformations to be portable as part of the VL JSON spec or dynamically updated by interactive selections, you will definitely want to use VL transformations rather than Arquero. Otherwise, using Arquero for your data transformations and then handing it to Vega or Vega-Lite is great, and in many cases can also end up being more efficient (faster, less memory usage).

Using Arquero with Vega and Vega-Lite is intended to be simple: you can simply pass an Arquero table as a data input. Given an iterable such as an Arquero table, the Vega runtime will get the iterated content. For Arquero tables this generates objects (just like the table `objects()` method) for each row.

However, be aware that if you are passing the same table as input to Vega/VL in multiple locations, you are incurring object serialization overhead each time. In those cases, you might want to instead call `table.objects()` once, and then pass in the resulting array of objects to all of your visualizations.

_Off-topic for Arquero_: Regarding restricted access URLs in Vega, you can potentially provide your own [vega-loader instance](https://github.com/vega/vega/tree/master/packages/vega-loader#loader). In particular, the `http` option can be used to provide options to `fetch`.
